syntax = "proto3";

package greet;

// The chat service definition.
service Chat {
    // Sends a message to a user
    rpc SendMessage (MessageSendRequest) returns (MessageSendResponse) {}

    // Open a stream of messages for a client; should never be called explicitly by user
    rpc ChatStream (ChatRequest) returns (stream SingleMessage) {}

    // Creates a new account
    rpc CreateAccount (CreateAccountRequest) returns (CreateAccountResponse) {}

    // Logs into an account
    rpc Login (LoginRequest) returns (LoginResponse) {}

    // Logs out of an account
    rpc Logout(LogoutRequest) returns (LogoutResponse) {}

    // Gets a list of accounts matching a wildcard
    rpc ListAccounts (ListAccountsRequest) returns (ListAccountsResponse) {}

    // Deletes an account
    rpc DeleteAccount (DeleteAccountRequest) returns (DeleteAccountResponse) {}

    // Gets server id
    rpc GetServerId (Empty) returns (GetServerIdResponse) {}

    // Gets primary server id
    rpc GetPrimaryServerId (Empty) returns (GetServerIdResponse) {}

    // Updates primary server
    rpc UpdatePrimaryServer (UpdatePrimaryServerRequest) returns (UpdatePrimaryServerResponse) {}

    // Server to server method: updates the chat length of a given user for the secondary servers
    rpc UpdateChatLength (UpdateChatLengthRequest) returns (UpdateChatLengthResponse) {}

    // Print all of the messages in the message queue for a given user
    rpc PrintMessages (PrintMessagesRequest) returns (PrintMessagesResponse) {}
}


// Request to send a message to a user
message MessageSendRequest {
    string recipient = 1;
    string sender = 2;
    string message = 3;
    bool fromPrimary = 4;
}

// Response for a message send
message MessageSendResponse {
    bool success = 1;
}

// Request to start a chat stream
message ChatRequest {
    string accountName = 1;
    bool fromPrimary = 2;
}

message SingleMessage {
    string sender = 1;
    string message = 2;
}

// Create a new account
message CreateAccountRequest {
    string accountName = 1;
    bool fromPrimary = 2;
}

// Response for a create account request
message CreateAccountResponse {
    bool success = 1;
}

// Log into an account using username
message LoginRequest {
    string accountName = 1;
    bool fromPrimary = 2;
}

// Response for a login request
message LoginResponse {
    bool success = 1;
}

// Log out of an account using username
message LogoutRequest {
    string accountName = 1;
    bool fromPrimary = 2;
}

// Response for a logout request
message LogoutResponse {
    bool success = 1;
}

// Request to get the list of accounts specified by a text wildcard
message ListAccountsRequest {
    string accountWildcard = 1;
}

// Response for a list accounts request
message ListAccountsResponse {
    repeated string accounts = 1;
}

// Request to delete an account
message DeleteAccountRequest {
    string accountName = 1;
    bool fromPrimary = 2;
}

// Response for a delete account request
message DeleteAccountResponse {
    bool success = 1;
}

// Empty message
message Empty {}

// Response for a get server id request
message GetServerIdResponse {
    int32 serverId = 1;
}

// Request to update primary server
message UpdatePrimaryServerRequest {
    int32 primaryServerId = 1;
}

// Response for an update primary server request
message UpdatePrimaryServerResponse {
    bool success = 1;
}

// Request to update chat length
message UpdateChatLengthRequest {
    string accountName = 1;
    int32 chatLength = 2;
}

// Response for an update chat length request
message UpdateChatLengthResponse {
    bool success = 1;
}

// Request to print messages
message PrintMessagesRequest {
    string accountName = 1;
}

// Response for a print messages request
message PrintMessagesResponse {
    repeated SingleMessage messages = 1;
}
